libraries/lib-uuid/Uuid.cpp

```diff
@@ -0,0 +1,216 @@
+/*!********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file Uuid.cpp
+ @brief Define a class to generate and parse UUIDs
+
+ Dmitry Vedenko
+ **********************************************************************/
+
+/*!********************************************************************
+
+ @class Uuid
+ @brief Platform independent class for generating and parsing UUIDs.
+
+ **********************************************************************/
+
+#include "Uuid.h"
+
+#include <algorithm>
+#include <cstring>
+
+#if defined(USE_UUID_CREATE)
+#   include <rpc.h>
+#elif defined(USE_CFUUID)
+#   include <CoreFoundation/CFUUID.h>
+#elif defined(USE_LIBUUID)
+#   include <uuid/uuid.h>
+#endif
+
+#include "HexHelpers.h"
+
+namespace audacity
+{
+
+namespace
+{
+bool readByte(Uuid::Bytes::iterator& outputIt,
+              std::string::const_iterator& inputIt,
+              const std::string::const_iterator& inputEnd)
+{
+   if (inputIt == inputEnd)
+      return false;
+
+   const char c1 = *inputIt++;
+
+   if (!std::isxdigit(c1))
+      return false;
+
+   if (inputIt == inputEnd)
+      return false;
+
+   const char c2 = *inputIt++;
+
+   if (!std::isxdigit(c2))
+      return false;
+
+   *outputIt = uint8_t((HexCharToNum(c1) << 4) | HexCharToNum(c2));
+
+   ++outputIt;
+
+   return true;
+}
+} // namespace
+
+Uuid::Uuid()
+    : Uuid(Bytes {})
+{
+}
+
+Uuid::Uuid(const Bytes& data) noexcept
```

Comment:
Interesting that you use noexcept.  I haven't formed that habit yet.  I often say in code comments that there are parts of the code that must be no-throw.  The annotation could be useful.

The compiler does not have to prove that exceptions are impossible from a noexcept function (and even if it could, the program would still be valid).  Rather the standard requires calls to terminate will happen if exceptions do try to escape.

But I'm not familiar with whatever warnings compilers make that could help you make such a proof.

Reply:
I will remove it here, it was just a force of habit. I worked on the platform with no exception, so having noexcept everywhere made code behave similarly everywhere

Reply:
I didn't say remove it!

Reply:
Well, it is wrong here anyway. `std::vector` has a throwable copy constructor. 

Anyway, besides the generated code size it only matters for move constructors. I will add one :)

Reply:
Oh, it has std::array, ignore my part about std::vector

Reply:
A separate move ctor wouldn't make sense for a std::array argument